File: test.spritebuilder/Source/TDEngine/Engine/MainEngine.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Engine;

import Map.LevelDTO;
import Map.LevelDTO.BadLevelInfo;
import Map.MapDAO;
import Unit.Solder.Solder;
import Unit.Solder.SolderFabric;
import Unit.Tower.Tower;
import Unit.Tower.TowerFabric;
import Utils.Point2D;
import internalObject.EngineOptions;
import internalObject.MapStateInside;
import internalObject.MapStateInternal;
import internalObject.TowerInternal;
import internalObject.TowerInternal.TowerType;
 
 

/**
 *
 * @author verS
 */
public class MainEngine {
     

    SolderEngine unitE;
    private TowerEngine towerEngine;
    SolderFabric unitFabric;
    TowerFabric towerFabric;
    LevelDTO levelInfo;
    MapDAO mapdao;
    PathEngine path;
    PlaceEngine place;
    MoneyEngine money;
    
   public static int colPassSolderToLoos = 10;
 
boolean isGameEnd=false;
    private boolean isGameWin;
    
   
  

    public MainEngine(MapDAO mapdao,SolderFabric unitFabric, TowerFabric towerFabric ) {
        this.mapdao = mapdao;
        this.unitFabric = unitFabric;
          this.towerFabric = towerFabric;
      
         
    }
    
    public void startLevel(LevelDTO levelInfo, int money) throws BadLevelInfo
    {  
         levelInfo.checked();
         this.money = new MoneyEngine(money);
       
          
         this.path = new PathEngine(levelInfo.path); 
         
         if(levelInfo.placesToSetTower != null)
            this.place = new PlaceEngine(levelInfo.placesToSetTower);
          if(levelInfo.placesToSetTowerNew != null) {
            this.place = new PlaceEngine(levelInfo.placesToSetTowerNew,(byte)0);
        }
         
         unitE = new SolderEngine(unitFabric,this.path,levelInfo.solders,this.money); 
          
         towerEngine = new TowerEngine(this.towerFabric,
                 this.place,unitE,this.money,this.path); 
         
        
    }
    
   
    public EngineEvent onGameEnd;

    public EngineEvent getOnGameEnd() {
        return onGameEnd;
    }

    public void setOnGameEnd(EngineEvent onGameEnd) {
        this.onGameEnd = onGameEnd;
    }


  
    
   
  
    
    
    public MapStateInternal getState()
    {
        return new MapStateInside(this.unitE.GetSoldersOnMap(),this.towerEngine.getTowers(),this.towerEngine.getShells(),
                
                
            this.isGameEnd,this.isGameWin,this.money.getMoney());
    }

    public void doStep() throws Exception {
        
     if(!isGameEnd){
        this.unitE.nextStep();
        this.towerEngine.nextStep();
        checkGameEnd();
        }
    }
    
    
    boolean checkWin()
    {
        
      return 
              this.unitE.isUnitEnd() && (!this.unitE.isSoldersOnMapNotDead());
      
    }
    
    boolean checkLoose()
    {
       return (this.unitE.getPassUnit() > MainEngine.colPassSolderToLoos);
    }
public void checkGameEnd()
{
    
    if(checkWin())
    {
             if(onGameEnd != null) {
            onGameEnd.execute();
        }
       isGameEnd = true;
       isGameWin = true;
    }
    
    if(checkLoose())
    {
      if(onGameEnd != null) {
            onGameEnd.execute();
        }
       isGameEnd = true;
       isGameWin = false;
    }
}

    public boolean setTower(TowerInternal.TowerType ID_Tower,Point2D  position) throws Exception {
        
        if(towerEngine == null) {
            throw new Exception("Level is not started. You need startLevel befor using setTower.");
        }
        return towerEngine.SetTower(ID_Tower,position);
    }

    public boolean canSetTower(TowerType towerType, Point2D position) {
        try {
            return towerEngine.canSetTower(towerType,position);
        } catch (Exception ex) {
            return false;
        }
    }

    public boolean isEnd() {
       return isGameEnd;
    }

    public void setOption(EngineOptions option) {
        Tower.engineOptions = option;
        Solder.engineOptions = option;
    }

    public void soldTower(int ID) {
        this.towerEngine.soldTower(ID);
    }

    

  
}

File: TowerDefence.spritebuilder/Engine/MainEngine.swift
//
//  MainEngine.swift
//  TowerDefence
//
//  Created by koveres on 08.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation

 
public class MainEngine {
    
    public struct colPassSolderToLoos{public static var value:CInt = 10;}
    
    var isGameEnd:Bool=false;
    var isGameWin:Bool=false;
    
    
    
    
    public init(  ) {
    
    
    }
    
    public func startLevel(    money:CInt,levelInfo:LevelDTO)
    {
         levelInfo.checked();
      EngAndFabric().resetEngineForLevel( money,levelInfo: levelInfo);
    }
    
    
    public var onGameEnd:EngineEvent?;
    
    public func getOnGameEnd() -> (EngineEvent?) {
        return onGameEnd;
    }
    
    public func setOnGameEnd( onGameEnd1: EngineEvent) {
        onGameEnd = onGameEnd1;
    }
    
    public func getState() -> (MapStateInternal)
    {
        return  MapStateInside(isEnd1: isGameEnd, isWin1: isGameWin);
    }
    
    public func doStep()  {
    
    if(!isGameEnd){
      EngAndFabric.inst_.soldeEng!.nextStep();
      EngAndFabric.inst_.towerEngine!.nextStep();
        checkGameEnd();
        }
    }
    
    
    func checkWin() -> (Bool)
    {
    
    return
     ( EngAndFabric.inst_.soldeEng!.isUnitEnd() && (!EngAndFabric.inst_.soldeEng!.isSoldersOnMapNotDead()) );
    
    }
    
    func checkLoose()-> (Bool)
    {
        return ( EngAndFabric.inst_.soldeEng!.getPassUnit() > MainEngine.colPassSolderToLoos.value);
    }
    public func checkGameEnd()
    {
    
    if(checkWin())
    {
    if(onGameEnd != nil) {
        onGameEnd!.execute();
    }
    isGameEnd = true;
    isGameWin = true;
    }
    
    if(checkLoose())
    {
    if(onGameEnd != nil) {
    onGameEnd!.execute();
    }
    isGameEnd = true;
    isGameWin = false;
    }
    }
    
    public func setTower(  ID_Tower:TowerType,   position:CGPoint) -> (Bool) {
    
    if(EngAndFabric.inst_.towerEngine == nil) {
     NSException(name: "Level is not started. You need startLevel befor using setTower.",reason: nil,userInfo: nil).raise();
    }
    return EngAndFabric.inst_.towerEngine!.SetTower(ID_Tower,position: position);
    }
    
    public func canSetTower( towerType:TowerType, position:CGPoint) -> (Bool) {
        return EngAndFabric.inst_.towerEngine!.canSetTower(towerType,position: position);
    
    }
    
    public   func isEnd() -> (Bool) {
    return isGameEnd;
    }
    
     
    
    public func soldTower(  ID:CInt) {
      EngAndFabric.inst_.towerEngine!.soldTower(ID);
    }
    
    
    
    
}
